OpenRA.Platforms.Default/Shader.cs

```diff
@@ -13,27 +13,32 @@
 using System.Collections.Generic;
 using System.IO;
 using System.Text;
+using OpenRA.Graphics;
 
 namespace OpenRA.Platforms.Default
 {
 	class Shader : ThreadAffine, IShader
 	{
-		public const int VertexPosAttributeIndex = 0;
-		public const int TexCoordAttributeIndex = 1;
-		public const int TexMetadataAttributeIndex = 2;
-		public const int TintAttributeIndex = 3;
-
 		readonly Dictionary<string, int> samplers = new Dictionary<string, int>();
 		readonly Dictionary<int, int> legacySizeUniforms = new Dictionary<int, int>();
 		readonly Dictionary<int, ITexture> textures = new Dictionary<int, ITexture>();
 		readonly Queue<int> unbindTextures = new Queue<int>();
 		readonly uint program;
+		readonly IShaderBindings bindings;
 
 		protected uint CompileShaderObject(int type, string name)
 		{
 			var ext = type == OpenGL.GL_VERTEX_SHADER ? "vert" : "frag";
-			var filename = Path.Combine(Platform.EngineDir, "glsl", name + "." + ext);
-			var code = File.ReadAllText(filename);
+			var filename = name + "." + ext;
+			string code;
+
+			if (Game.ModData != null && Game.ModData.DefaultFileSystem.TryOpen(filename, out var stream))
```

Comment:
As above: we should define the shaders as string properties(?) on the specific subclasses instead of doing gymnastics to load them from files.

Reply:
The idea of having them as files is the platform. If you have a different platform you might use hlsl shaders for example. Hardcoding them would remove the posibility to provide multiple shader languages.

Reply:
Many other things would also need to change if we wanted to switch from OpenGL to DirectX, etc, so this doesn't really change things by itself. If we did want to support multiple then IMO that is even more reason to have everything grouped in one place so that the modder doesn't just implement things for DirectX/Windows and ignore the others.

Reply:
Another benefit of this approach is that it lets the renderer do arbitrary modifications to the shader source before compiling it, whereas right now any new `#ifdef`s etc would need to be hardcoded into `Shader.cs`.

We already have a usecase: `combined.frag` writing to `gl_FragDepth` forces the GPU to turn off certain hardware level optimizations for all mods, even though most of them dont use the depth buffer feature. Wrapping this in an `#ifdef` that we only enable for mods with `EnableDepthBuffer: True` gives a ~10% performance improvement to everything else. This is important for the performance constrained raspberry pi and remastered mods, so is something that I would like to implement for the next release.